docs: map coordinate systems (#1561)
vaadin/docs: articles/components/map/index.asciidoc

See the <<Feature Flag>> section for how to do this.
====

> .Breaking Changes Since 23.1

Author: As soon as we merge this, all examples using coordinates will become invalid for developers using 23.1, and multiple sections refer to the default coordinate system being EPSG:4326.
I was aware that adding another blob to the top of the page is not ideal, but in this case I think it's appropriate. We can not assume that developers read the whole page, including the breaking changes or coordinate system section, before they start trying things.
I would not do this if we had a separate deployment for the next version, or 23.2 was already released, but as long as 23.1 is the default, we should add as many hints about this as possible IMO.

> .Breaking Changes Since 23.1

Author: I think I would be more relaxed about this if devs could make the code examples work out of the box. Following your suggestion to address the issue in the relevant places, how about adding code comments to all the places where coordinates are created, that indicate that you should use `Coordinate.fromLonLat` when on 23.1?

> .Breaking Changes Since 23.1

Reviewer: That's a great idea IMO

> .Breaking Changes Since 23.1

Author: Added comments to the code examples.

== Coordinate Systems

The default coordinate system, or projection, for all coordinates passed to, or received from the component is EPSG:4326, also referred to as GPS coordinates or latitude and longitude.
This is called the user projection, as it affects all the coordinates that the user of the component, in this case the developer, is working with.
That means when setting coordinates, for example of the viewport or for a marker, coordinates must be specified in EPSG:4326 by default, and coordinates received from events are guaranteed to be in that projection.

Internally, the component converts all coordinates into the projection that is used by the map's view, which is called the view projection.
The default view projection is EPSG:3857, also known as https://en.wikipedia.org/wiki/Web_Mercator_projection[Web Mercator Projection^].

=== Changing the User Projection

The user projection can be changed, for example when working with coordinates that are not in the EPSG:4326 projection.
Out of the box the component only has support for the EPSG:4326 and EPSG:3857 projections, however the component allows <<#defining-custom-projections, defining custom projections>>.

> As mentioned above, changing the user projection affects *all* coordinates, using coordinates with different projections is not allowed.

Author: Done

.Example for setting a different user projection
[.example]
--
[source,Java]
----
// Set user projection to EPSG:3857
Map.setUserProjection("EPSG:3857");

// Create a map, center viewport on New York using
// coordinates in EPSG:3857 projection
Map map = new Map();
map.setCenter(new Coordinate(-8235375.352932, 4967773.457877));
map.setZoom(10);
----
--

.User Projection Scope
[NOTE]
====
The user projection setting is valid for the lifetime of the current browser page, which means it stays defined when navigating using the router, but not when doing a hard location change, or reloading the page.

> In practice, we recommend setting the user projection on every page where it is used.

Author: Done

The user projection affects all maps on the browser page, and should be set before creating any maps.

> Changing the user projection does not affect existing maps, specifically coordinates configured in an existing map are not automatically converted into the new projection.

Reviewer: This admonition feels a bit verbose. Perhaps it's enough to end it with

> "The user projection affects all maps on the browser page, and should be set before creating any maps."

Instead, maps should be recreated after changing this setting.
====

=== Defining Custom Projections

> Custom projections can be defined when working with coordinates that are neither in the EPSG:4326 nor EPSG:3857 projection.

Author: Done

> Defining a custom projection requires a name, which can later be referenced, as well as a projection definition in the Well Known Text (WKT) format.

Reviewer: should we link to https://www.iso.org/standard/76496.html for the WKT standard?

    + "        AUTHORITY[\"EPSG\",\"4258\"]],\n"
    + "    PROJECTION[\"Transverse_Mercator\"],\n"
    + "    PARAMETER[\"latitude_of_origin\",0],\n"
    + "    PARAMETER[\"central_meridian\",27],\n"
    + "    PARAMETER[\"scale_factor\",0.9996],\n"
    + "    PARAMETER[\"false_easting\",500000],\n"
    + "    PARAMETER[\"false_northing\",0],\n"
    + "    UNIT[\"metre\",1,\n"
    + "        AUTHORITY[\"EPSG\",\"9001\"]],\n"
    + "    AXIS[\"Easting\",EAST],\n"
    + "    AXIS[\"Northing\",NORTH],\n"
    + "    AUTHORITY[\"EPSG\",\"3067\"]]";
Map.defineProjection("EPSG:3067", wkt);

// Set user projection to EPSG:3067
Map.setUserProjection("EPSG:3067");

// Create a map, change view projection to EPSG:3067 as well
Map map = new Map();
map.setView(new View("EPSG:3067"));

// Center viewport on Helsinki using
// coordinates in EPSG:3067 projection
map.setCenter(new Coordinate(385725.63, 6671616.89));
map.setZoom(10);
----
--

.Custom Projection Scope
[NOTE]
====
Custom projections are valid for the lifetime of the current browser page, which means they stay defined when navigating using the router, but not when doing a hard location change, or reloading the page.

> In practice, we recommend defining projections on every page where they are used.

Reviewer: ```suggestion
In practice, it is advisable to define projections on every page where they are used.
```

Author: Done

[NOTE]
====
Custom projections are valid for the lifetime of the current browser page, which means they stay defined when navigating using the router, but not when doing a hard location change, or reloading the page.
In practice, it is advisable to define projections on every page where they are used.

Custom projections must be defined before using them as either user projection or view projection.
====

=== Breaking Changes Since 23.1

In 23.2, the default coordinate system has been changed to EPSG:4326, also referred to as GPS coordinates, or latitude and longitude.
As this documentation has already been updated to the 23.2, all documentation sections refer to the new coordinate system, and examples have been adapted to create coordinates using latitude and longitude.

If you are still using 23.1, note that the default coordinate system is EPSG:3857, and all coordinates provided to the map must be in that projection.
If you want to create coordinates from latitude and longitude in 23.1, use the static `Coordinate.fromLonLat(longitude, latitude)` helper.
You can easily adapt the examples from this documentation to work with 23.1 by changing code that constructs coordinates from:

> ```java
> new Coordinate(longitude, latitude)
> ```
> to:
> ```java
> Coordinate.fromLonLat(longitude, latitude)
> ```

Reviewer 2: ```suggestion
[source,java,role="before"]
----
new Coordinate(longitude, latitude)
----

[source,java,role="after"]
----
Coordinate.fromLonLat(longitude, latitude)
----
```

2. Add the following content: `com.vaadin.experimental.mapComponent=true`
3. Restart the application.

> == Breaking Changes Since 23.1

Reviewer: The contents of this section could also be a lot more compact, since the Coordinate Systems section already describes the default projection etc. I would just have a single paragraph describing how 23.1 differs from that, and advise to use the fromLonLat method to adapt the examples.

> == Breaking Changes Since 23.1

Author: Similar reason as above, having this on the second level makes it more discoverable, as it shows up in the TOC. If someone runs into a problem that their markers do not show up because they are using EPSG:3857 coordinates, I'm not sure if the coordinate system section is the first place to look.

I don't think it's too elaborate. It provides context for a problem that developers might run into, and provides the solution, rather than readers have to puzzle everything together from different sections. Maintenance of this section is not a problem IMO, as we'll remove it rather soon.

> == Breaking Changes Since 23.1

Reviewer: Fair enough, but surely we cannot possibly need _both_ the big admonition at top _and_ a top-level section about it, so if we keep the admonition, I'd still move this under Cooridinates, since we can be pretty sure they've gotten the message already.
